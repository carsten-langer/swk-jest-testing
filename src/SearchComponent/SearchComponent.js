import debounce from 'debounce-promise';

const SEARCH_RESPONSE_EVENT_NAME = 'searchComponent:response';
const SEARCH_ERROR_EVENT_NAME = 'searchComponent:error';

export default class SearchComponent {

    constructor(searchClient) {
        this.debouncedSearch = debounce(searchClient.search, 200, { accumulate: true });
    }

    search(query) {
        return this.debouncedSearch(query)
            .then(this.emitSearchResponseEvent)
            .catch(this.emitSearchErrorEvent);
    }

    emitSearchResponseEvent(searchResponse) {
        console.log('search response', searchResponse);

        document.dispatchEvent(new CustomEvent(SEARCH_RESPONSE_EVENT_NAME, {
            detail: searchResponse
        }));
    }

    emitSearchErrorEvent(searchError) {
        document.dispatchEvent(new CustomEvent(SEARCH_ERROR_EVENT_NAME, {
            detail: searchError
        }));
    }
}