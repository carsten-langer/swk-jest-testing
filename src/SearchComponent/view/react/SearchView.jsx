import React from 'react';

export default ({ searchComponent, placeholderText }) => {
  const searchEventHandler = ({ key }) => {
    searchComponent.search(key);
  };

  return (
    <input
      className="search-box"
      type="search"
      placeholder={placeholderText}
      onKeyDown={searchEventHandler}
    />
  );
};