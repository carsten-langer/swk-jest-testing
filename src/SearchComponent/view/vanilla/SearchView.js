export default class SearchView {
    constructor(searchComponent, placeholderText) {
        this.searchComponent = searchComponent;
        this.placeholderText = placeholderText;
    }

    searchEventHandler({ key }) {
        this.searchComponent.search(key);
    }

    render() {
        const searchViewElement = document.createElement('input');
        searchViewElement.setAttribute('type', 'search');
        searchViewElement.setAttribute('placeholder', this.placeholderText);
        searchViewElement.classList.add('search-box');
        searchViewElement.addEventListener('keydown', this.searchEventHandler.bind(this));

        return searchViewElement;
    }
};