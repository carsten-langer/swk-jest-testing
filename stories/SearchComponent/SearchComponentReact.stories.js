// import React from 'react';
// import fetchMock from 'fetch-mock';

// import ReactSearchView from '../../src/SearchComponent/view/react/SearchView';
// import SearchComponent from '../../src/SearchComponent/SearchComponent';
// import SearchClient from '../../src/SearchClient/SearchClient';

// export default { title: 'SearchComponent' };

// export const asReactComponent = () => {
//   const searchComponent = new SearchComponent(new SearchClient());
//   const placeholder = 'some placehodler text';
//   const props = { searchComponent, placeholder };
//   const searchView = (<ReactSearchView {...props} />);

//   document.addEventListener('searchComponent:response', (response) => console.log('success', response.detail));
//   document.addEventListener('searchComponent:error', (response) => console.log('error', response.detail));

//   return searchView;
// };