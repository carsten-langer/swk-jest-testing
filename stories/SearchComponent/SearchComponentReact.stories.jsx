import React from 'react';
import fetchMock from 'fetch-mock';

import ReactSearchView from '../../src/SearchComponent/view/react/SearchView';
import SearchComponent from '../../src/SearchComponent/SearchComponent';
import SearchClient from '../../src/SearchClient/SearchClient';

fetchMock.get('begin:/api/search?query=', {
  body: [
    'google.com',
    'googol.de',
    'mail.google.com'
  ]
}, { status: 200 });

export default { title: 'SearchComponent' };

export const asReactComponent = () => {
  const searchComponent = new SearchComponent(new SearchClient());
  const placeholderText = 'some placeholder text';
  const props = { searchComponent, placeholderText };
  const searchView = (<ReactSearchView {...props} />);

  document.addEventListener('searchComponent:response', (response) => console.log('success', response.detail));
  document.addEventListener('searchComponent:error', (response) => console.log('error', response.detail));

  return searchView;
};