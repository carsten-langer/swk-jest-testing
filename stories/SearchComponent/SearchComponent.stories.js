import fetchMock from 'fetch-mock';

import SearchView from '../../src/SearchComponent/view/vanilla/SearchView';
import SearchComponent from '../../src/SearchComponent/SearchComponent';
import SearchClient from '../../src/SearchClient/SearchClient';

fetchMock.get('begin:/api/search?query=', {
  body: [
    'google.com',
    'googol.de',
    'mail.google.com'
  ]
}, { status: 200 });

export default { title: 'SearchComponent' };

export const withCustomPlaceholder = () => {
  const searchView = new SearchView(new SearchComponent(new SearchClient()), 'please enter your search...');

  document.addEventListener('searchComponent:response', (response) => console.log('success', response.detail));
  document.addEventListener('searchComponent:error', (response) => console.log('error', response.detail));

  return searchView.render()
};