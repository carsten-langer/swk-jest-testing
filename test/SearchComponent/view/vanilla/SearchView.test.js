import SearchComponent from '../../../../src/SearchComponent/SearchComponent';
import SearchView from '../../../../src/SearchComponent/view/vanilla/SearchView';

jest.mock('../../../../src/SearchComponent/SearchComponent');

describe('Vanilla javascript search view', () => {

    beforeEach(() => {
        document.body.innerHTML = '<div id="searchBox" />';
    });

    it('should render a search box', () => {
        const { viewContainer, searchViewElement } = renderSearchView();

        expect(searchViewElement).not.toBeNull();
        expect(viewContainer.innerHTML).toMatchSnapshot();
    });

    it('should search when user enters an input', () => {
        const { searchViewElement, searchComponent } = renderSearchView();

        searchViewElement.dispatchEvent(new KeyboardEvent('keydown', {
            key: 'g'
        }));

        expect(searchComponent.search).toHaveBeenCalledWith('g');
    });

    function renderSearchView() {
        const searchComponent = new SearchComponent();
        const searchView = new SearchView(searchComponent, 'some placehodler text');
        const viewContainer = document.getElementById('searchBox');

        searchView.render(viewContainer);

        return {
            searchComponent,
            viewContainer,
            searchViewElement: viewContainer.querySelector('.search-box')
        };
    }
});