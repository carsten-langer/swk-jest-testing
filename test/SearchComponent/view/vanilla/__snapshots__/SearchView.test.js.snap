// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vanilla javascript search view should render a search box 1`] = `"<input type=\\"search\\" placeholder=\\"some placeholder text\\" class=\\"search-box\\">"`;
